### PR TITLE
fix: parsing of basic auth remote-url

### PR DESCRIPTION
**Pull request by ANGkeith**
not too sure how to fix the code smell though

possible fix for #1178 and #921

### CODE REVIEW

**Comment by firecow** [comment]
https://sonarcloud.io/dashboard?id=firecow_gitlab-ci-local&pullRequest=1202 Have a look at the code smells @ANGkeith 

I'll mark the security findings as safe.

**Comment by ANGkeith** [comment]
> Have a look at the code smells @ANGkeith

yeah, i am saw those code smells, not too sure how to fix it : x 


some of the named groups are used, but the sonarqube seemed to fail to detect it.

**Comment by firecow** [comment]
> > Have a look at the code smells @ANGkeith
> 
> yeah, i am aware of those code smells, not too sure how to fix it : x
> 
> They (named groups) are used, but the sonarqube seemed to fail to detect it.

Oh... Then we shall ignore. I didn't think SonarQube could produce false positives. Never experienced that.

**Comment by firecow** [comment]
@ANGkeith I've fixed the code smells with https://github.com/firecow/gitlab-ci-local/pull/1202/commits/6f89d910437cc27036d5e8a1c88ef637bf6a1d95

File: src/git-data.ts [hunk 1]
```diff
@@ -82,24 +82,44 @@ export class GitData {
...
         try {
-            const {stdout: gitRemote} = await Utils.spawn(["git", "remote", "-v"], cwd);
-            const gitRemoteMatch = /(?<schema>git|https?)(?::\/\/|@)(?<host>[^:/]*)(:(?<port>\d+)\/|:|\/)(?<group>.*)\/(?<project>[^ .]+)(?:\.git)?.*/.exec(gitRemote);
-
-            assert(gitRemoteMatch?.groups != null, "git remote -v didn't provide valid matches");
-
-            this.remote.schema = gitRemoteMatch.groups.schema;
-            if (this.remote.schema === "git") {
+            let gitRemoteMatch;
+            const {stdout: gitRemote} = await Utils.spawn(["bash", "-c", "git remote get-url gcl-origin 2> /dev/null || git remote get-url origin"], cwd);
```
Comment (firecow) [inline src/git-data.ts:86]: Can you explain the logic behind this?

Comment (ANGkeith) [reply src/git-data.ts:86]: > Can you explain the logic behind this?
https://github.com/firecow/gitlab-ci-local/issues/921#issuecomment-2090224041
`gcl-origin` is to for people to customize their own "remote" in the event that `origin` is not what they want

Comment (ANGkeith) [reply src/git-data.ts:86]: > If bash is really needed Utils.bash should be used.
> Keep in mind also, that bash is a lot slower, than just calling plain ol' shell cmd's

Ok, got it, wasnt aware of this.

Comment (ANGkeith) [reply src/git-data.ts:86]: > Wouldn't this be cleaner?
LGTM

**Comment by ANGkeith** [comment]
> @ANGkeith I've fixed the code smells with [6f89d91](https://github.com/firecow/gitlab-ci-local/commit/6f89d910437cc27036d5e8a1c88ef637bf6a1d95)

thanks!

File: src/git-data.ts [hunk 2]
```diff
@@ -82,24 +82,44 @@ export class GitData {
...
+                this.remote.host = gitRemoteMatch.groups.host;
+                this.remote.group = gitRemoteMatch.groups.group;
+                this.remote.project = gitRemoteMatch.groups.project;
+                this.remote.schema = gitRemoteMatch.groups.schema;
                 this.remote.port = gitRemoteMatch.groups.port ?? "22";
+            } else {
+                gitRemoteMatch = /(?<username>\S+)@(?<host>[^:]+):(?<group>\S+)\/(?<project>[^ .]+)\.git/.exec(gitRemote); // regexr.com/7vjoq
+                assert(gitRemoteMatch?.groups != null, "git remote get-url gcl-origin 2> /dev/null || git remote get-url origin");
+
+                const {stdout: port} = await Utils.bash(`ssh -G ${gitRemoteMatch.groups.username}@${gitRemoteMatch.groups.host} | grep '^port' | awk '{print $2}'`);
```
Comment (ANGkeith) [inline src/git-data.ts:116]: btw, for the fix for #1083

This should set the correct port number to use for 
https://github.com/firecow/gitlab-ci-local/blob/250bab2232a65f89f6813ebc67d99fe2b5b00dbf/src/parser-includes.ts#L203

File: src/git-data.ts (hunk 1 above)
Comment (ANGkeith) [reply src/git-data.ts:86]: > Can you explain the logic behind this?
> 
> ```ts
> let gitRemote;
> try {
>     const res = await Utils.spawn(["git", "remote", "get-url", "gcl-origin"], cwd);
>     gitRemote = res.stdout;
> } catch {
>     const res = await Utils.spawn(["git", "remote", "get-url", "origin"], cwd);
>     gitRemote = res.stdout;
> }
> ```
> 
> Wouldn't this be cleaner? (I don't like bash magic) 😄

yeah lgtm,though i am a bash magician 😂

Comment (firecow) [reply src/git-data.ts]: Haha, yeah... But it ramps up, spawning that many bash instances :smile:

**Review by firecow — APPROVED** [review]

**Pull request merged**